app/Factory/ProductAllocationFactory.php

```diff
@@ -0,0 +1,37 @@
+<?php
+/**
+ * Invoice Ninja (https://invoiceninja.com).
+ *
+ * @link https://github.com/invoiceninja/invoiceninja source repository
+ *
+ * @copyright Copyright (c) 2024. Invoice Ninja LLC (https://invoiceninja.com)
+ *
+ * @license https://www.elastic.co/licensing/elastic-license
+ */
+
+namespace App\Factory;
+
+use App\Models\ProductAllocation;
+
+class ProductAllocationFactory
+{
+    public static function create(int $company_id, int $user_id, int $product_id): ProductAllocation
```

Comment:
linking with the product_id is risky. we do not use product IDs anywhere in the application as a relationship.

We use product_key to link products to invoice items... however, there is no unique constraint on the product_key, so resolving duplicates becomes an issue.

I am not averse to using the product id cc @hillelcoren your thoughts appreciated.

Reply:
my take on this is, that the product_allocation table could mainly be seen as an extened data to the product table itself.
We can definatly add the product_key into the request objects to determine the product_id if you like.

Within the invoice the product_key is still used as before.

But I am open for advice, how to implement this table, as you suggest :)